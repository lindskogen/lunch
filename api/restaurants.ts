import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../lib/lunch-fetcher/fetch-restaurants";

export default async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};
