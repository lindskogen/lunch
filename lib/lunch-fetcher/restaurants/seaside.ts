import * as _ from "lodash";
import { isWeekday, isNotNull, isNotZero } from "../lib/utils";

export const url = "http://seaside.kvartersmenyn.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $(".day .meny")
    .toArray()
    .map(item => {
      const allNodes = item.childNodes
        .filter(i => i.tagName === "strong" || i.nodeValue)
        .map(item => $(item).text());

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const foodsRegex = /^([A-Ã–]+( [0-9])?) ?[\n :]([^0-9:\n]+)/gm;

          const allFoodsForWeekDay = foods.join("\n");

          const parsedFoods = [];

          let res = foodsRegex.exec(allFoodsForWeekDay);
          while (res !== null) {
            const [, title, , name] = res;
            parsedFoods.push({ title, name: name.trim() });
            res = foodsRegex.exec(allFoodsForWeekDay);
          }

          return {
            wday: wday as WeekDay,
            items: parsedFoods
          };
        });
    });

  return { name: "Seaside", url, days: _.flatten(menus) };
};
