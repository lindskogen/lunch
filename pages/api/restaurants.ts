import { NowRequest, NowResponse } from "@now/node";
import { fetchAll } from "../../lib/lunch-fetcher/fetch-restaurants";

const restaurants = async (req: NowRequest, res: NowResponse) => {
  const restaurants = await fetchAll();

  res.setHeader("Cache-Control", "s-maxage=3600, stale-while-revalidate");

  res.json({ restaurants, lastUpdated: new Date().toISOString() });
};

export default restaurants;
