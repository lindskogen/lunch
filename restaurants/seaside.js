const _ = require("lodash");
const { isWeekday, isNotNull, isNotZero } = require("../lib/utils");

const url = "http://seaside.kvartersmenyn.se/";
module.exports.url = url;

module.exports.parseHtml = $ => {
  const menus = $(".day .meny")
    .toArray()
    .map(item => {
      const allNodes = item.childNodes
        .filter(i => i.tagName === "strong" || i.nodeValue)
        .map(item => $(item).text());

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const parsedFoods = foods.map(food => {
            const res = /^([A-Ã–]+( [0-9])?) ?:?([^0-9]*)/.exec(food);
            if (res) {
              const [all, title, number, name] = res;
              return { title, name: name.trim() };
            } else {
              return null;
            }
          });

          return {
            wday,
            items: parsedFoods
          };
        });
    });

  return { name: "Seaside", url, days: _.flatten(menus) };
};
