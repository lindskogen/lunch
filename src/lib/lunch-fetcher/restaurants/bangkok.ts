import { Item, RestaurantData } from "./bangkok_types";
import { groupBy, take } from "lodash-es";
import { filterMap, mapDayNameToWeekDay } from "../lib/utils";
import { RestaurantDayMenu, FoodItem, Restaurant } from "../types";

const QUERY = `
query Query($clientId: String!) {
  items(clientId: $clientId) {
    id
    name
    preview
    info
    description
    outOfStock
    sortOrder
    category
    ingestions {
        price
        type
    }
    itemCategory
  }
  itemCategories(clientId: $clientId) {
      id
      name
      preview
      description
      category
  }
}
`;

const CLIENT_ID = "5c0644cbfd30240bfdb26765";

export const url = "https://www.bkgbg.se/lunch";

export const fetchMeals = (): Promise<RestaurantData> =>
  fetch("https://secure.paidit.se/graphql", {
    method: "POST",
    headers: {
      "content-type": "application/json",
    },
    body: JSON.stringify({
      query: QUERY,
      variables: { clientId: CLIENT_ID },
      operationName: "Query",
    }),
  })
    .then((r) => r.json())
    .then((r) => r.data);

const transformToFoodItem = (item: Item): FoodItem => {
  const isVegetarian = item.description.match(/VEGETARISK/);
  const isVegan = item.description.match(/VEGAN/);

  return {
    name: item.description
      .replace(/\(?VEGETARISK\)? ?/, "")
      .replace(/\(?VEGAN\)? ?/, "")
      .replace(/ \./g, "."),
    title: (
      item.name.replace(/,$/g, "").replace(",", " — ") +
      (isVegetarian ? " — VEGETARISK" : isVegan ? " — VEGANSK" : "")
    ).toUpperCase(),
  };
};

const groupMeals = (data: RestaurantData): RestaurantDayMenu[] => {
  const itemsByItemCategory = groupBy(data.items, (item) => item.itemCategory);

  const days = data.itemCategories.filter((category) =>
    category.name.toLowerCase().includes("lunch")
  );

  return filterMap(days, (day) => {
    const dayName = mapDayNameToWeekDay(day.name);

    if (!dayName) {
      return null;
    }

    return {
      items: take(itemsByItemCategory[day.id], 5).map(transformToFoodItem),
      wday: dayName,
    };
  });
};

export const parseResponse = (restaurantData: RestaurantData): Restaurant => ({
  name: "Bangkok Kitchen",
  url,
  days: groupMeals(restaurantData),
});
