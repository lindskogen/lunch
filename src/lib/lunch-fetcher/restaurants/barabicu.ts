import { filterMap } from "../lib/utils";
import { Restaurant, RestaurantDayMenu, WeekDay } from "../types";

export const url = "https://barabicu.se/";

const weekdayTranslations: WeekDay[] = [
  WeekDay.Monday,
  WeekDay.Tuesday,
  WeekDay.Wednesday,
  WeekDay.Thursday,
  WeekDay.Friday,
];
export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#todays-lunch li")
    .toArray()
    .map<RestaurantDayMenu>((dayNode, index) => {
      return {
        items: filterMap($("h3", dayNode).toArray(), (elem) => {
          if ($(elem).text() === "" || $(elem.next).text() === "") {
            return null;
          }
          return {
            name: $(elem.next).text(),
            title: $(elem).text().split("â€¢")[0].trim(),
          };
        }),
        wday: weekdayTranslations[index],
      };
    });

  return { name: "Barabicu", url, days: menus };
};
