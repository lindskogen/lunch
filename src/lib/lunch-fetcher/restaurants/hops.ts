import { flatMap, includes, takeWhile, flatten } from "lodash-es";
import { id, isNotNull, isNotZero, isWeekday } from "../lib/utils";
import { FoodItem, Restaurant, WeekDay } from "../types";

export const url = "http://www.hopsbar.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $("#Veckansmeny")
    .toArray()
    .map(item => {
      const allNodes = flatMap(item.childNodes, node =>
        $(node)
          .text()
          .trim()
          .split(/\n/)
      )
        .map(s => s.trim())
        .filter(id);

      const weekMealsStartIndex = allNodes.findIndex(
        text => text === "Veckans rätter"
      );

      const weekItems: FoodItem[] = takeWhile(
        allNodes.slice(weekMealsStartIndex + 1),
        text => !includes(text, " kaffe ")
      ).map(food => {
        const foodname = food.trim();
        const matches = foodname.split(':');

        if (matches.length > 1) {
          const [title, name] = matches;
          return { name: name.trim(), title: title.trim().toUpperCase() };
        }

        return { name: foodname };
      });

      const splitIndexes = allNodes
        .map((text, index) => (isWeekday(text) ? index : null))
        .filter(isNotNull);

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map(splitIndex => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return days
        .filter(([wday]) => isWeekday(wday))
        .map(([wday, ...foods]) => {
          const items = foods.map(food => ({ name: food.trim() }));

          return {
            wday: wday as WeekDay,
            items: takeWhile(
              items,
              ({ name }) => name !== "Veckans rätter"
            ).concat(weekItems)
          };
        });
    });

  return { name: "Hops", url, days: flatten(menus) };
};
