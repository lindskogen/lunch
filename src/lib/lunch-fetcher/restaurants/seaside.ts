import { flatten } from "lodash-es";
import {
  filterMap,
  isNotZero,
  isWeekday,
  mapDayNameToWeekDay,
} from "../lib/utils";
import { Restaurant, WeekDay } from "../types";

export const url = "http://seaside.kvartersmenyn.se/";

export const parseHtml = ($: CheerioStatic): Restaurant => {
  const menus = $(".day .meny")
    .toArray()
    .map((item) => {
      const allNodes = item.childNodes
        .filter((i) => i.tagName === "strong" || i.nodeValue)
        .map((item) => $(item).text());

      const splitIndexes = filterMap(allNodes, (text, index) =>
        isWeekday(text) ? index : null
      );

      let lastIndex = 0;
      const days = splitIndexes
        .filter(isNotZero)
        .map((splitIndex) => {
          const slice = allNodes.slice(lastIndex, splitIndex);
          lastIndex = splitIndex;
          return slice;
        })
        .concat([allNodes.slice(lastIndex)]);

      return filterMap(days, ([wday, ...foods]) => {
        const weekday = mapDayNameToWeekDay(wday);

        if (!weekday) {
          return null;
        }

        const foodsRegex = /^([A-Ã–]+( [0-9])?) ?[\n :]([^0-9:\n]+)/gm;

        const allFoodsForWeekDay = foods.join("\n");

        const parsedFoods = [];

        let res = foodsRegex.exec(allFoodsForWeekDay);
        while (res !== null) {
          const [, title, , name] = res;
          parsedFoods.push({ title, name: name.trim() });
          res = foodsRegex.exec(allFoodsForWeekDay);
        }

        return {
          wday: weekday,
          items: parsedFoods,
        };
      });
    });

  return { name: "Seaside", url, days: flatten(menus) };
};
