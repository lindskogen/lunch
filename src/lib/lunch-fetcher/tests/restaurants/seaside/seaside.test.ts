import * as cheerio from "cheerio";
import { parseHtml } from "../../../restaurants/seaside";

import { mock as mock1 } from "./mock1";
import { mock as mock2 } from "./mock2";

describe("seaside", () => {
  describe("when provided with mock 1", () => {
    const result = parseHtml(cheerio.load(mock1));
    it("should work", () => {
      expect(result).toMatchSnapshot();
    });

    it("should have 3 foods per day", () => {
      result.days.forEach((day) => {
        expect(day.items).toHaveLength(3);
      });
    });

    it("should parse 5 days", () => {
      expect(result.days).toHaveLength(5);
    });

    it("should remove colon from title and name", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.title).not.toMatch(/:/);
          expect(item.name).not.toMatch(/:/);
        });
      });
    });

    it("should have no empty names", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.name).not.toHaveLength(0);
        });
      });
    });
  });
  describe("when provided with mock 2", () => {
    const result = parseHtml(cheerio.load(mock2));
    it("should work", () => {
      expect(result).toMatchSnapshot();
    });

    it("should have 6 foods per day", () => {
      result.days.forEach((day) => {
        expect(day.items).toHaveLength(6);
      });
    });

    it("should parse 5 days", () => {
      expect(result.days).toHaveLength(5);
    });

    it("should remove colon from title and name", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.title).not.toMatch(/:/);
          expect(item.name).not.toMatch(/:/);
        });
      });
    });

    it("should have no empty names", () => {
      result.days.forEach((day) => {
        day.items.forEach((item) => {
          expect(item.name).not.toHaveLength(0);
        });
      });
    });
  });
});
