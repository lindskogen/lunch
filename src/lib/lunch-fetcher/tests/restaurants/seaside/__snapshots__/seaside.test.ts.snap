// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seaside when provided with mock 1 should work 1`] = `
Object {
  "days": Array [
    Object {
      "items": Array [
        Object {
          "name": "Grillad andouillekorv serveras med sötpotatis och chiliaioli samt rostad potatis och grönsaker",
          "title": "GÅRD",
        },
        Object {
          "name": "Fiskgryta på kummel, torsk, lax, blåmusslor och rotfrukter serveras med potatisstomp",
          "title": "HAV",
        },
        Object {
          "name": "Tofu Tikka masala serveras med naanbröd, mangochutney och basmatiris",
          "title": "ÄNG",
        },
      ],
      "wday": "Måndag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Linguine med nattbakad högrev, basilika och soltorkad tomatsås samt riven parmesan",
          "title": "GÅRD",
        },
        Object {
          "name": "Örtstekt kummelfilé serveras med dill och räksås, kokt potatis och blomkålspuré",
          "title": "HAV",
        },
        Object {
          "name": "Tofu Tikka masala serveras med naanbröd, mangochutney och basmatiris",
          "title": "ÄNG",
        },
      ],
      "wday": "Tisdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Ostgratinerat kycklingbröst serveras med basilikasås och pilaffris samt ruccola",
          "title": "GÅRD",
        },
        Object {
          "name": "Pankopanerad sejfilé serveras med gochujangmajonnäs, rostad sesam och jasminris",
          "title": "HAV",
        },
        Object {
          "name": "Rödbetsbiff med fetaost serveras med rostad potatis och persiljekräm på färskost",
          "title": "ÄNG",
        },
      ],
      "wday": "Onsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Pannbiff serveras med löksky, persilja, rårörda lingon, saltgurka och potatispuré",
          "title": "GÅRD",
        },
        Object {
          "name": "Potatis och purjolökssoppa med krutonger och bacon samt pannkakor med sylt och grädde",
          "title": "SOPPA",
        },
        Object {
          "name": "Rödbetsbiff med fetaost serveras med rostad potatis och persiljekräm på färskost",
          "title": "ÄNG",
        },
      ],
      "wday": "Torsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Pulled porkburger serveras med BBQ-sås, coleslaw, picklad rödlök och friterad potatis",
          "title": "GÅRD",
        },
        Object {
          "name": "Ugnsbakad torskfilé serveras med citronsmörsås, haricots verts och gratinerad potatis",
          "title": "HAV",
        },
        Object {
          "name": "Ricotta och spenatfylld tortellini serveras med ratatouille och riven ost",
          "title": "ÄNG",
        },
      ],
      "wday": "Fredag",
    },
  ],
  "name": "Seaside",
  "url": "http://seaside.kvartersmenyn.se/",
}
`;

exports[`seaside when provided with mock 2 should work 1`] = `
Object {
  "days": Array [
    Object {
      "items": Array [
        Object {
          "name": "Strimlad biff och bambuskott i soyasås med rättikasallad serveras med ris",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Gräddstuvad pytt i panna med stekt ägg och inlagda rödbetor",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Marockansk rotfruktsgryta med couscous samt aubergine & valnötsröra (vegan)",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein & plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv & bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Måndag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Stekt fläsk med bruna bönor eller löksås samt kokt potatis",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Fiskgryta med bouillabaisesås, rotfrukter och fänkål samt aioli",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Chili sin carne serveras med crème fraiche, picklad rödlök och råris",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Tisdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Kycklingfilé med medelhavsgrönsaker,  aubergine & bönröra samt örtbulgur",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Panerad rödspätta med curryremouladesås, citron & kokt potatis",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Chili sin carne serveras med crème fraiche, picklad rödlök och råris",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Onsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Ärtsoppa med fläsk och köttkorv samt pannkakor med sylt och vispad grädde",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Husets lasagne serveras med  tomat, ruccola och rödlökssallad",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Lins och böngryta med sambal, koriander och kokosmjölk serveras med ris (vegan)",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Torsdag",
    },
    Object {
      "items": Array [
        Object {
          "name": "Wallenbergare med potatispuré, gröna ärtor, rårörda lingon & brynt smör",
          "title": "DAGENS 1",
        },
        Object {
          "name": "Örtbakad sejfilé med lökfräst bladspenat, dillslungad potatis och smörsås",
          "title": "DAGENS 2",
        },
        Object {
          "name": "Dagens pasta",
          "title": "PASTA",
        },
        Object {
          "name": "Lins och böngryta med sambal, koriander och kokosmjölk serveras med ris (vegan)",
          "title": "VEGETARISK",
        },
        Object {
          "name": "Välj ditt eget protein och plocka resten ifrån vår stora salladsbuffé",
          "title": "SALLAD",
        },
        Object {
          "name": "Ägg, korv och bacon med stekt potatis & stekt ägg",
          "title": "GRILL",
        },
      ],
      "wday": "Fredag",
    },
  ],
  "name": "Seaside",
  "url": "http://seaside.kvartersmenyn.se/",
}
`;
